nemo/utils/exp_manager.py

```diff
@@ -0,0 +1,213 @@
+from typing import List, Optional, Union, Dict
+
+from omegaconf import OmegaConf, DictConfig
+from hydra.utils import get_original_cwd
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from nemo.constants import NEMO_ENV_VARNAME_DATETIME
+from nemo.utils import logging
+from nemo.utils.get_rank import is_global_rank_zero
+
+
+@dataclass
+class ExpManagerConfig:
+    """TODO: Is this how you document a dataclass?
+    Args:
+        root_dir (str, Path): The base directory to create the logging directory. Defaults to None, which logs to
+            ./NeMo_experiments.
+        name (str): The name of the experiment. Defaults to None, which uses the lightning default of "default".
+        create_tensorboard_logger (bool): Whether to create a tensorboard logger and attach it to the pytorch lightning
+            trainer. Defaults to True.
+        create_checkpoint_callback (bool): Whether to create a ModelCheckpoint callback and attach it to the pytorch
+            lightning trainer. The ModelCheckpoint saves the top 3 models with the best "val_loss" as well as the most
+            recent model. Defaults to True.
+        files_to_copy (list): A list of files to copy to the experiment logging directory. Defaults to None which copies
+            no files.
+    """
+
+    root_dir: Optional[str] = None
+    name: Optional[str] = None
+    create_tensorboard_logger: Optional[bool] = True
+    create_checkpoint_callback: Optional[bool] = True
+    files_to_copy: Optional[List[str]] = None
+
+
+def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictConfig, Dict]] = None):
+    """
+    exp_manager is a helper function used to manage folders for experiments. It follows the pytorch lightning paradigm
+    of root_dir/model_or_experiment_name/version. If the lightning trainer has a logger, exp_manager will get root_dir,
+    name, and version from the logger. Otherwise it will use the root_dir and name arguments to create the logging
+    directory. The version will be a datetime string if running single node, and version will be an integer if running
+    on slurm multi-node.
+    It optionally creates TensorBoardLogger, and ModelCheckpoint objects from pytorch lightning. It copies sys.argv,
+    and git information if available to the logging directory. It creates a log file for each process to log their
+    output into.
+
+    Args:
+        trainer (pytorch_lightning.Trainer): The lightning trainer.
+        cfg (ExpManagerConfig)
+    """
+    schema = OmegaConf.structured(ExpManagerConfig)
+    if cfg is None:
+        cfg = schema
+    elif isinstance(cfg, dict):
+        cfg = OmegaConf.create(cfg)
+    elif not isinstance(cfg, DictConfig):
+        raise ValueError(f"cfg was type: {type(cfg)}. Expected either a dict or a DictConfig")
+    # Ensure passed cfg is compliant with ExpManagerConfig
+    cfg = OmegaConf.merge(schema, cfg)
```

Comment:
This simply merged the two configs, it does not perform an actual test to see if LHS config is superset of RHS config. If user misspells and passes some random args, your defaults will be applied instead of raising an error below (since the keys would not match). I suggest not to perform this merge. If the cfg is None, simply assume user doesn't want exp manager and return.

Reply:
This is a valid point.
Regarding the mistake of misspelling, it seems easier to make `name` a MISSING / required argument in this case.

>  If the cfg is None, simply assume user doesn't want exp manager, log it and return.
This seems very hard to do incorrectly since the user has to explicitly call exp_manager()

> it does not perform an actual test to see if LHS config is superset of RHS config.
Alternatively, we could do
```python
cfg = OmegaConf.merge(schema, cfg)
OmegaConf.merge(cfg, schema)
```

Reply:
Actually the double-sided merge will not catch any errors. Let me take another look

Reply:
Yes name: missing is a good default. 

As to the default of returning on None, my thought process was we could add this rxp_manager to all scripts like in old Nemo by default, and the user cfg would decide if that would actually be used or not. If the config exists, the function call already exists so directly you get exp manager support. If you skip this in config then no exp manager support. 

The first merge there would augment cfg with the new values from the schema, and the second merge would be a basic no-op. All values already exist in the union. Think of merge as a nested union more than an intersection for OmegaConf behaviour.

Reply:
Actually it does catch typos, in the config if I change `name: *name` to `namw: *name`, OmegaConf raises the following error: `omegaconf.errors.ConfigKeyError: Key 'namw' not in 'ExpManagerConfig'`
Is this not what we want? Are there other issues that might arise?

Reply:
So the only thing left is the optional null of cfg and the name: ??? default

Reply:
done